new perf blogpost fix1 (#2463)
OpenLiberty/blogs: posts/2022-08-08-daytrader8-performance.adoc

// change the "IMAGE CAPTION" to a couple words of what the image is
// // // // // // // //

In this blog post, I will be discussing a performance comparison that demonstrates Liberty’s capabilities as the premier runtime for Java/Jakarta EE applications. The metrics used to perform this comparison, when running the sample Java EE8 application DayTrader8 are:
* First response - time from issuing the 'server start' command to the first response received by client 
* Memory footprint - Java process size (RSS) after first response is processed 
* Throughput - rate at which requests are processed 
* Scalability - throughput as the number of clients increases 

> The remainder of this blog post provides more detail about how the evaluation was conducted on four application servers that implement the Java EE8 specification and the results seen for each metric.

Reviewer: ```suggestion
We ran a sample Jakarta EE 8 application, on the four runtimes that implement the Jakarta EE 8 specification, including Open Liberty. Open Liberty outperformed the other application servers significantly, including starting up almost 80% faster than the nearest competitor, consuming less than half the memory consumed by the runtime of the nearest competitor, and scaling up to higher throughout than the competitors while simultaneously using less memory during load.
```

> The DayTrader8 application simulates an online stock trading system, allowing users to perform several actions to manipulate their portfolio. DayTrader8 is written in a manner that exercises several different features in the Java EE8 specification, making it a good representative benchmark application for evaluating different implementations. In the competitive comparison, we used the container images that each application server product had released, thus ensuring that we used the out-of-the-box settings for each application server. Note that the container images used JDK 11 for all the application servers, and HotSpot-based Open JDK was used for all the frameworks except Liberty for which the OpenJ9-based IBM Semeru JDK was used since this is used as the base JDK image for building the Liberty image. For Liberty, we used the steps outlined in the documentation to build the application container image, and we used Liberty’s powerful “composability” feature to configure only the required features used by the application. All the runs were done on a Linux SLES 12.4 system, running on a Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz machine, with containers limited to 4 CPUs and 4GB of RAM. 

Reviewer: ```suggestion
The sample application we used, link:https://github.com/OpenLiberty/sample.daytrader8[DayTrader8], simulates an online stock trading system, allowing users to perform several actions to manipulate their portfolio. DayTrader8 makes use of several different features in the Jakarta EE 8 specification, making it a good representative benchmark application for evaluating different implementations. For details of how we ran the benchmark tests, see <<method, Method>>.
```

> The first metric we compared was the time to receive the first response back at the client (sometimes equated with startup time). Liberty was almost 80% better than the nearest competitor (JBoss EAP) and this was a testament not just to the EE8 application server technology in Liberty but also to the optimized containers that include a shared classes cache population step as part of the container build procedure.    

Reviewer: Add sub-headings to break it up.

```suggestion
== Nearly 80% faster startup time

The first metric we compared was the time to receive the first response back at the client (sometimes equated with startup time). Liberty was almost 80% better than the nearest competitor (JBoss EAP) and this was a testament not just to the EE8 application server technology in Liberty but also to the optimized containers that include a shared classes cache population step as part of the container build procedure.    
```

//-------------------------
//[.img_border_light]
image::img/blog/dt8pic1.png[First Response Times ,width=70%,align="center"]
//-------------------------

> The next metric compared was the memory footprint just after the first response was sent back to the client. This shows the memory consumed by the server framework itself, separate from the memory that will be used by the application as workload increases. Liberty takes less than half the memory as the nearest competitor, while still consistently outperforming on the other performance metrics.  

Reviewer: ```suggestion
== Less than half the memory than the nearest competitor

The next metric compared was the memory footprint just after the first response was sent back to the client. This shows the memory consumed by the server framework itself, separate from the memory that will be used by the application as workload increases. Liberty takes less than half the memory as the nearest competitor, while still consistently outperforming on the other performance metrics.  
```

 
//-------------------------
//[.img_border_light]
image::img/blog/dt8pic2.png[Memory Footprint ,width=70%,align="center"]
//-------------------------

> Throughput is usually the metric many customers associate with “performance” since it measures the number of requests that the application server was able to handle in a fixed unit of time. As the below comparison shows, Liberty has a very well tuned Java EE8 implementation designed for both optimal peak throughput performance and judicious resource use. 

Reviewer: ```suggestion
== Higher throughput than competitors

Throughput is usually the metric many customers associate with “performance” since it measures the number of requests that the application server was able to handle in a fixed unit of time. As the graph shows, Liberty has higher throughput than the other runtimes. Liberty's very well-tuned Jakarta  EE 8 implementation is designed for both optimal peak throughput performance and judicious resource use. 
```

  
//-------------------------
//[.img_border_light]
image::img/blog/dt8pic3.png[Throughput ,width=70%,align="center"]
//-------------------------

> Finally, the last metric that we evaluated was the scalability (sometime also referred to as “scaling up”, since there is only a single server instance). In this evaluation we varied the number of simultaneous client requests (we were using JMeter to drive the load) so that we could get a view into the application server’s behavior at different load levels to assess if the above throughput differences were unique to a particular configuration. The scalability experiments used 2 CPUs overall and so it was relatively easy to saturate the CPU with a moderate number of client requests as can be seen by the throughput curves flattening out in the below comparison. Once again Liberty stands out for the superior throughput at many different load levels showing that it can scale up consistently as load increases (harder to judge at the lower load levels since it is likely that CPU was not saturated).

Reviewer: Re " (harder to judge at the lower load levels since it is likely that CPU was not saturated)" - I've removed it because it doesn't seem to add much, slightly contradicts the previous statement, and softens the finding itself.

```suggestion
== Scales up with load increases to higher throughput than competitors

The final metric that we evaluated was "scaling up", the scalability of the server instance. In this evaluation we varied the number of simultaneous client requests (we were using JMeter to drive the load) so that we could get a view into the runtime’s behavior at different load levels to assess if the previously described throughput differences were unique to a particular configuration.

The scalability experiments used two CPUs overall, so it was relatively easy to saturate the CPU with a moderate number of client requests as can be seen by the throughput curves flattening out in the following chart. Once again Liberty stands out for the superior throughput at many different load levels showing that it can scale up consistently as load increases.
```

//-------------------------
//[.img_border_light]
image::img/blog/dt8pic4.png[Throughput ,width=70%,align="center"]
//-------------------------

> To get a view into physical memory use (RSS) during load, we also plotted the ratio between throughput and memory at the different load levels to provide a view into how much “throughput per byte of memory” is being delivered by the different application servers. The fact that this last graph shows an even larger advantage for Liberty means that it was able to deliver higher throughput while simultaneously taking lesser memory during load. 

Reviewer: ```suggestion
== Less memory with higher throughput during load

To get a view into physical memory use (RSS) during load, we also plotted the ratio between throughput and memory at the different load levels to provide a view into how much “throughput per byte of memory” is being delivered by the different runtimes. This last graph shows an even larger advantage for Liberty which was able to deliver higher throughput while simultaneously taking less memory during load. 
```

 
//-------------------------
//[.img_border_light]
image::img/blog/dt8pic5.png[Throughput/Memory ,width=70%,align="center"]
//-------------------------

> In conclusion, we evaluated the performance of different application servers running a Java EE8 workload in containers, and the Liberty application server came out on top across the board in all metrics, showing how it is the most performant and cost-effective option that can help users save money when running Java/Jakarta EE workloads. We encourage you to try out Liberty for your application to experience these performance benefits, and we would love to hear any feedback that you might have as well! 

Reviewer: ```suggestion
== Conclusion

We evaluated the performance of different Jakarta EE 8 runtime workloads in containers, and Open Liberty came out on top in all metrics, showing how it is the most performant and cost-effective option that can  link:{url-prefix}/blog/2020/10/21/memory-footprint-throughput.html[help users save money] when running Java EE and Jakarta EE workloads.

We encourage you to try out Liberty for your application to experience these performance benefits, and we would love to hear any feedback that you might have as well! 

If you're new to Open Liberty, try out our link:{url-prefix}/guides/getting-started.html[Getting Started guide]. Or find out more about link:{url-prefix}/blog/2022/06/13/mpJWT-performance-improvements.html[how we doubled throughput of microservices secured with JWT].
```

> 

Reviewer: Insert new section at the end (like we did for the CRIU post):

**Also add a link to the "documentation" for building the container image as it's not clear where that refers to.**

```suggestion
[#method]
== Method
The sample application that we used for the benchmark tests, link:https://github.com/OpenLiberty/sample.daytrader8[DayTrader8], makes use of several different features in the Jakarta EE 8 specification, making it a good representative benchmark application for evaluating different implementations.

In the competitive comparison, we used the container images that each runtime had link:https://jakarta.ee/compatibility/download/[released for the Jakarta EE 8 Platform], which ensured that we used the out-of-the-box settings for each runtime. The container images used JDK 11 for all the runtimes, and HotSpot-based Open JDK for all the runtimes except Open Liberty which uses the OpenJ9-based IBM Semeru JDK in the base JDK layer of the Liberty image. For Liberty, we used the steps outlined in the documentation to build the application container image, and made use of Liberty's composability by configuring only the required features used by the application. All the runs were done on a Linux SLES 12.4 system, running on a Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz machine, with containers limited to 4 CPUs and 4GB of RAM. 
```
